### PR TITLE
[hypre] add the 'headers' property

### PR DESCRIPTION
**Pull request by v-dobrev**
... plus a small tweak in the 'libs' property.

### CODE REVIEW

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -110,8 +110,20 @@ def install(self, spec, prefix):
...
+        """Export the main hypre header, HYPRE.h; all other headers can be found
+        in the same directory.
+        Sample usage: spec['hypre'].headers.cpp_flags
+        """
+        hdrs = HeaderList(find(self.prefix.include, 'HYPRE.h', recursive=False))
+        return hdrs or None
+
     @property
     def libs(self):
-        is_shared = self.spec.satisfies('+shared')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/hypre/package.py]: I don’t remember details, but this and the new way are not always the same.

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 2]
```diff
@@ -110,8 +110,20 @@ def install(self, spec, prefix):
...
-        return find_libraries('libHYPRE', root=self.prefix,
-                              shared=is_shared, recursive=True)
+        """Export the hypre library.
+        Sample usage: spec['hypre'].libs.ld_flags
+        """
+        libs = find_libraries('libHYPRE', root=self.prefix.lib,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/hypre/package.py]: Are they always in ‘lib’ and not ‘lib64’ on some platforms?

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (v-dobrev) [reply var/spack/repos/builtin/packages/hypre/package.py]: Which one is the correct (or just preferred)?

Comment (davydden) [reply var/spack/repos/builtin/packages/hypre/package.py]: @alalazo probably knows the difference 😄

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 2 above)
Comment (v-dobrev) [reply var/spack/repos/builtin/packages/hypre/package.py]: I'll add fallbacks to search `prefix.lib64` and as a last resort all of `prefix`.

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hypre/package.py]: @v-dobrev Reading this discussion right now. I think @davydden refers to #4135

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 3]
```diff
@@ -110,8 +110,27 @@ def install(self, spec, prefix):
...
     def libs(self):
-        is_shared = self.spec.satisfies('+shared')
-        return find_libraries('libHYPRE', root=self.prefix,
-                              shared=is_shared, recursive=True)
+        """Export the hypre library.
+        Sample usage: spec['hypre'].libs.ld_flags
+        """
+        search_paths = [[self.prefix.lib, False], [self.prefix.lib64, False],
+                        [self.prefix, True]]
+        is_shared = '+shared' in self.spec
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hypre/package.py]: I also use this a lot, and I think it's fine. But, if we have to be subtle,  this means that at least one spec in the DAG as the `+shared` variant.

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (davydden) [reply var/spack/repos/builtin/packages/hypre/package.py]: yes, thanks for the link, @alalazo . I rembembered that there was a discussion somewhere

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 3 above)
Comment (v-dobrev) [reply var/spack/repos/builtin/packages/hypre/package.py]: Actually, in the current implementation, this is not true. I tested this with some debug prints: in `Spec.__contains__` the string `'+shared'` is turned (by the `_autospec` call) into a `Spec`: `'hypre+shared'` (printed as a string). So, even though the dependencies are traversed, they will not satisfy the spec `'hypre+shared'`.

Anyway, for me, this confirms that `'+shared' in self.spec` has the right behavior.

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 4]
```diff
@@ -110,8 +110,27 @@ def install(self, spec, prefix):
                         '-rhsone')
             make("install")
 
+    @property
+    def headers(self):
+        """Export the main hypre header, HYPRE.h; all other headers can be found
+        in the same directory.
+        Sample usage: spec['hypre'].headers.cpp_flags
+        """
+        hdrs = HeaderList(find(self.prefix.include, 'HYPRE.h',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/hypre/package.py]: I would just use `find_headers` instead.

Comment (v-dobrev) [reply var/spack/repos/builtin/packages/hypre/package.py]: OK, I'll switch to `find_headers`.

**Pull request merged**